read-fonts/src/tables/gsub/closure.rs

```diff
@@ -99,52 +235,43 @@ impl Gsub<'_> {
         {
             lookup_ids.extend(feature?.lookup_list_indices().iter().map(|idx| idx.get()));
         }
-
-        // and now we need to add lookups referenced by contextual lookups,
-        // IFF they are reachable via the current set of glyphs:
-        for lookup in lookup_list.lookups().iter() {
-            let subtables = lookup?.subtables()?;
-            match subtables {
-                SubstitutionSubtables::Contextual(tables) => tables
-                    .iter()
-                    .try_for_each(|t| t?.add_reachable_lookups(glyphs, &mut lookup_ids)),
-                SubstitutionSubtables::ChainContextual(tables) => tables
-                    .iter()
-                    .try_for_each(|t| t?.add_reachable_lookups(glyphs, &mut lookup_ids)),
-                _ => Ok(()),
-            }?;
-        }
         Ok(lookup_ids)
     }
 }
 
+impl GlyphClosure for SubstitutionLookup<'_> {
+    fn add_reachable_glyphs(&self, ctx: &mut ClosureCtx) -> Result<(), ReadError> {
+        self.subtables()?.add_reachable_glyphs(ctx)
+    }
+}
+
 impl GlyphClosure for SubstitutionSubtables<'_> {
-    fn add_reachable_glyphs(&self, glyphs: &mut IntSet<GlyphId16>) -> Result<(), ReadError> {
+    fn add_reachable_glyphs(&self, glyphs: &mut ClosureCtx<'_>) -> Result<(), ReadError> {
         match self {
             SubstitutionSubtables::Single(tables) => tables.add_reachable_glyphs(glyphs),
             SubstitutionSubtables::Multiple(tables) => tables.add_reachable_glyphs(glyphs),
             SubstitutionSubtables::Alternate(tables) => tables.add_reachable_glyphs(glyphs),
             SubstitutionSubtables::Ligature(tables) => tables.add_reachable_glyphs(glyphs),
             SubstitutionSubtables::Reverse(tables) => tables.add_reachable_glyphs(glyphs),
-            _ => Ok(()),
+            SubstitutionSubtables::Contextual(tables) => tables.add_reachable_glyphs(glyphs),
+            SubstitutionSubtables::ChainContextual(tables) => tables.add_reachable_glyphs(glyphs),
         }
     }
 }
 
 impl<'a, T: FontRead<'a> + GlyphClosure + 'a, Ext: ExtensionLookup<'a, T> + 'a> GlyphClosure
     for Subtables<'a, T, Ext>
 {
-    fn add_reachable_glyphs(&self, glyphs: &mut IntSet<GlyphId16>) -> Result<(), ReadError> {
-        self.iter()
-            .try_for_each(|t| t?.add_reachable_glyphs(glyphs))
+    fn add_reachable_glyphs(&self, ctx: &mut ClosureCtx<'_>) -> Result<(), ReadError> {
+        self.iter().try_for_each(|t| t?.add_reachable_glyphs(ctx))
     }
 }
 
 impl GlyphClosure for SingleSubst<'_> {
-    fn add_reachable_glyphs(&self, glyphs: &mut IntSet<GlyphId16>) -> Result<(), ReadError> {
+    fn add_reachable_glyphs(&self, ctx: &mut ClosureCtx<'_>) -> Result<(), ReadError> {
         for (target, replacement) in self.iter_subs()? {
```

Comment:
Same as Harfbuzz, we could apply some optimization here: https://github.com/harfbuzz/harfbuzz/blob/d3e4977c708837d32b6d6e8f2ed4ea3045bce09a/src/OT/Layout/GSUB/SingleSubstFormat2.hh#L42